### PR TITLE
[rpc] modify test_local_shutdown_with_rpc to not be flaky

### PR DESCRIPTION
**Pull request by rohan-varma**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#30640 [rpc] modify test_local_shutdown_with_rpc to not be flaky**

This test would get very occasional flakes, with an error saying the
RPC timed out. This happened because one worker would still be waiting for the
return value of an RPC, but another worker had already performed its local
shutdown, so it would not have sent the response. This didn't show up in
initial testing since the flakiness is very rare (< 1/100 test runs). We can add a dist.barrier() that all processes call into to alleviate the issue.

Differential Revision: [D18775731](https://our.internmc.facebook.com/intern/diff/D18775731/)

### CODE REVIEW

File: test/rpc_test.py [hunk 1]
```diff
@@ -1193,17 +1177,20 @@ def test_local_shutdown_with_rpc(self):
...
+        #  RPCs will be completed, so we catch timeout errors here.
+        try:
+            rpc.rpc_sync(
+                "worker{}".format(dst_rank),
+                torch.add,
+                args=(torch.ones(n, n), torch.ones(n, n)),
+            )
+        except RuntimeError as e:
+            self.assertTrue("RPC ran for more than" in str(e))
         # pass in graceful=False to ensure that we don't wait for other workers.
```
Comment (pritamdamania87) [inline test/rpc_test.py]: Can't we just add a `dist.barrier` before the shutdown to fix the problem?

Comment (rohan-varma) [reply test/rpc_test.py]: Good idea, I think this is a better solution.

File: test/rpc_test.py [hunk 2]
```diff
@@ -1132,15 +1116,22 @@ def test_local_shutdown_with_rpc(self):
...
+        # For `RpcAgent` other than `ProcessGroupAgent`,
+        # no `_default_pg` is initialized.
+        if not dist.is_initialized():
+            dist.init_process_group(
+                backend="gloo",
+                init_method=self.init_method,
+                rank=self.rank,
+                world_size=self.world_size,
+            )
+        dist.barrier()
```
Comment (xush6528) [inline test/rpc_test.py]: I think it's easier to understand than an ad-hoc helper function, `_check_rpc_done`, put in this file.

Still not too perfect to initialize process group. I think later, if `_wai_all_workers()` supports calling multiple times. We could use it.

**Review by xush6528 — APPROVED** [review]

**Review by zhaojuanmao — APPROVED** [review]

**Comment by rohan-varma** [comment]
Closing in favor of https://github.com/pytorch/pytorch/pull/30837, since this is having rebase errors according to github.

**Pull request closed without merging**